src/device-state.coffee

```diff
@@ -406,6 +410,37 @@ module.exports = class DeviceState extends EventEmitter
 			apps = _.keyBy(_.map(appsArray, singleToMulticontainerApp), 'appId')
 			return { apps, config: deviceConf }
 
+	restoreBackup: (targetState) =>
+		@setTarget(targetState)
+		.then =>
+			appId = _.keys(targetState.local.apps)[0]
+			if !appId?
+				throw new Error('No appId in target state')
+			volumes = targetState.local.apps[appId].volumes
+			backupPath = path.join(constants.rootMountPoint, 'mnt/data/backup')
+			rimraf(backupPath) # We clear this path in case it exists from an incomplete run of this function
+			.then ->
+				mkdirp(backupPath)
+			.then ->
+				execAsync("tar -xzf backup.tgz -C #{backupPath} .", cwd: path.join(constants.rootMountPoint, 'mnt/data'))
+			.then ->
+				fs.readdirAsync(backupPath)
+			.then (dirContents) =>
+				Promise.mapSeries dirContents, (volumeName) =>
+					if volumes[volumeName]?
+						console.log("Creating volume #{volumeName} from backup")
+						# If the volume exists (from a previous incomplete run of this restoreBackup), we delete it first
+						@applications.volumes.get({ appId, name: volumeName })
+						.then =>
+							@applications.volumes.remove({ appId, name: volumeName })
+						.catch(NotFoundError, _.noop)
+						.then =>
+							@applications.volumes.createFromPath({ appId, name: volumeName, config: volumes[volumeName] }, path.join(backupPath, volumeName))
```

Comment:
I assume this is the part of the code which confused @samothx too? E.g. the volume didn't exist in the target state but it did in the backup - Might be worth printing the volumes which we're ignoring as well, so it's more transparent? This way we could tell the difference between say a corrupted backup and incorrect target state.

Reply:
True... now I've changed it so that it errors out if anything in the backup file is unexpected (either it has something that's not a directory, or it has a directory that's not a valid volume in the target state). What do you think?